KEP-5295: KYAML (#5296)
kubernetes/enhancements: keps/sig-cli/5295-kyaml/README.md

Author: ACK

Always quote strings, including ALL map keys and field names
<<[/UNRESOLVED>>

Slices and arrays are rendered with square-braces. Elements are rendered as per this list.

Pointers are rendered as `null` or dereferenced and rendered as per the pointee’s type.

Types which implement the Go standard `json.Marshaler` interface (`MarshalJSON()`) are rendered by marshaling to JSON, unmarshaling into an `any`, and then rendering that `any` as KYAML.

Types which implement the `yaml.Marshaler` interface (`MarshalYAML()`) are rendered as per their type.  `MarshalYAML()` is not a Go standard and is not used by KYAML.

<<[UNRESOLVED yaml self-marshaling]>>
Should we do something for YAML self-marshalling?  It is non-standard and we do not use it in Kubernetes.
<<[/UNRESOLVED>>

Interfaces are rendered as `null` or their true value type.

Any other type (as per `reflect.Value.Kind()`) will produce an error.

### Multi-line strings

For better or worse, multi-line strings are a thing that people do in Kubernetes, and conventional YAML makes it pretty easy. KYAML has to choose how we want to handle it.  Consider these examples of conventional YAML.

> Example 1:

Reviewer: I think we neglected to mention that the spec doesn't permit this string format when in block style (braces etc), only flow style (indentation sensitive yaml).

Therefore these are not an option, they're incompatible with formatting in block style.

Author: Will note

Example 2:
```
example_2: |
    this:
      is:
      - embedded
      - yaml
      where:
        whitespace:
          matters: true
```

#### Option 1. Treat it like a Go quoted-and-escaped string

This always comes out as one line.

```
example_1: “This\nis a\nmulti-line string\n”
example_2: "this:\n  is:\n  - embedded\n  - yaml\n  where:\n    whitespace:\n      matters: true\n"
```

Pro: Easy and well-defined
Pro: Can be pasted into most C-like languages to get the original string.
Con: Hard to read and write by humans when it gets long.

#### Option 2. Use YAML flow-folding and escaped line-breaks

This gets closer to the original without sacrificing too much of the goals of this KEP.

```
example_1: ”\
    This\n\
    is a\n\
    multi-line string\
    “
example_2: "\
    this:\n\
    \  is:\n\
    \  - embedded\n\
    \  - yaml\n\
    \  where:\n\
    \    whitespace:\n\
    \      matters: true\n\
    "
```

Pro: Easier to read than option 1 and close to the original
Pro: Closer to the original than option 3
Con: Requires weird EOL handling (`\n\` meaning a newline, plus a line-continuation which  prevents the fold from creating a single leading space on the next line)

> Con: Subtle handling of leading whitespace (all dropped unless escaped)

Reviewer: Worth noting: This is true of all multi-line yaml strings *except* option 1), even in the flow string styles we can't use.

#### Option 3: Use YAML flow-folding and double line-breaks

This also gets close to the original without sacrificing too much of the goals of this KEP.

```
example_1: ”\
    This

    is a

    multi-line string\
    ”
example_2: "\
    this:

    \  is:

    \  - embedded

    \  - yaml

    \  where:

    \    whitespace:

    \      matters: true

    "
```

Pro: Easier to read than option 1 and close to the original
Pro: Less escapes than option 2

> Con: Requires weird and surprising EOL handling (double-newline being a directive to YAML to add a single newline to the output)

Reviewer: We should've also mentioned the surprising behavior of adding a single leading space if you drop one of these lines:

IE:
```yaml
this: "
  is

  - embedded
  "
```
== `this: "is\n- embedded"`
While:
```yaml
this: "
  is
  - embedded
  "
```
==  `this: "is\n - embedded` (notice the space)

I think this is a really bizarre footgun.

We should revisit this specification when that happens.

### Risks and Mitigations

Risk: Our encoder produces some output which is not valid YAML.
Mitigation: That’s a bug. It’s hard to imagine a case that can’t be fixed.

Risk: The result could be net worse than plain YAML.
Mitigation: We will produce many examples and include them in the review process. We will solicit feedback from various groups on the result.

Risk: This gives a false sense of safety, but doesn’t SOLVE  the problems with YAML, just avoids them.
Mitigation: Docs?

Risk: Like all APIs, people will come to depend on the details, which may prevent us from changing them.
Mitigation: Be as precise as possible where we are confident, but leave room where we are not.

## Design Details

Creating an encoder for this format is not very technically challenging,
but there are many details to consider.

### Packaging

<<[UNRESOLVED where it lives]>>

The simplest thing to do is to write a new implementation of k8s.io/cli-runtime/pkg/printers.ResourcePrinter (in staging/src/k8s.io/cli-runtime/pkg/printers). However, this isn’t the most obviously reusable place for it.

> A more ecosystem-friendly solution is to write a new function or package under  sigs.k8s.io/yaml](sigs.k8s.io/yaml) which can be used by other tools.

Reviewer 2: It would be nicer to provide the encoder in a module with no dependencies on anything outside of the standard library. We discussed using KYAML as pretty-printer for logging and test output. A dependency on the encoder will be easier to justify if it does not pull in other dependencies.

Alternatively, is there a way to avoid the [github.com/google/go-cmp and gopkg.in/check.v1 dependencies](https://github.com/kubernetes-sigs/yaml/blob/0fe7da35edc6aa625b56143f278741913b204464/go.mod#L6-L7), for example by moving tests into a separate module? Presumably they are only used for testing, but a consumer of the module cannot be sure.

Reviewer: > Alternatively, is there a way to avoid the [github.com/google/go-cmp and gopkg.in/check.v1 dependencies](https://github.com/kubernetes-sigs/yaml/blob/0fe7da35edc6aa625b56143f278741913b204464/go.mod#L6-L7), for example by moving tests into a separate module? Presumably they are only used for testing, but a consumer of the module cannot be sure.

We already have these in sigs.k8s.io/yaml by way of the go-yaml forks. They are test-only so they are part of your module resolution graph but not part of your actual dependency set.

Author: Depending on the decisions around yamlfmt, it's easier or harder to modularize.  If I have to write all the logic to encode/decode/encode, I am loathe to do it twice.

That said, maybe I am wrong?  The pure encoder for kyaml is only about 600 LOC and some of that could probably go away if we don't need to staisfy the PrintObj interface.

Reviewer 2: I'm thinking of code which does not currently import `sigs.k8s.io/yaml`.

> They are test-only so they are part of your module resolution graph but not part of your actual dependency set.

I know that, but dependency analysis tools don't :wink:

Reviewer: > I'm thinking of code which does not currently import sigs.k8s.io/yaml.

If that code wants to import Kubernetes style YAML formatting, I think it's reasonable to have the same requirements as sigs.k8s.io/yaml

> I know that, but dependency analysis tools don't 😉

Eh? They only don't for resolving the graph, it shouldn't show up in your own go.mod / vendor.

Do we have a concrete blocker here, other than "it would be nice"? I agree that having none is nice, but it may be impractical, and I don't think the KEP is the right place to determine the dependencies of the implementing code.

Reviewer 2: Our own dependency tracking in Kubernetes is graph-based. It doesn't matter what a dependency is used for (test or non-test code), if an unwanted dependency shows up, it gets flagged as a problem. I'm sure other projects do the same.

This is not entirely theoretic. I would like to move `k8s.io/kubernetes/test/utils/ktesting` into a staging repo. We talked about that a while ago, I just had to put it on hold. Currently, that ktesting package depends on sigs.k8s.io/yaml as a pretty-printer. I doubt that sigs.k8s.io/yaml and thus github.com/google/go-cmp + gopkg.in/check.v1 would be accepted as new dependency for some of our staging repos. I could switch to KYAML and if the KYAML encoder was in a module with no other dependencies, then it would be easier to argue for allowing it as new dependency.

Reviewer 2: My other use case is as an on-demand pretty-printer in klog (`klog.Dump(obj any)`), with a `klog.Diff(old, new any)` as replacement for `diff := cmp.Diff(old, new); klog.Info(...., "diff", diff)`.
- Would we be okay to have a simple `sigs.k8s.io/kyaml` with the output code as dependency of klog? Maybe.
- Would `sigs.k8s.io/yaml` as it stands today be okay? I doubt it.

/cc @dims

Author: I see some value in having a encoder which is really stand-alone, and I don't think I would object to having 2 implementations as long as we have solid tests?

Reviewer: > I doubt that sigs.k8s.io/yaml and thus github.com/google/go-cmp + gopkg.in/check.v1 would be accepted as new dependency for some of our staging repos. 

I'm not sure about that. sigs.k8s.io/yaml is already a dependency of many staging libraries.
Anything using client-go imports it via client-go

Which staging repo imports klog (and therefore has runtime logic not just simple types) yet doesn't use a client?

Author: Worse, since we are parsing to `yaml.Node`, kyaml depends on `"sigs.k8s.io/yaml/goyaml.v3"` anyway.

Both sigs.k8s.io/yaml/go-yaml.v2 and v3 offer a way to decode YAML into an AST, which we can render as KYAML.  Neither, however, offers a way to encode an arbitrary Go object as the equivalent YAML AST.  To do that we may have to encode the object to plain YAML, then decode it to AST, then render it as KYAML.  That extra conversion is ugly, but this is not generally performance-sensitive code and is not run on the server.

[https://github.com/goccy/go-yaml](https://github.com/goccy/go-yaml) offers an API which appears to do what we need in both directions.

### Header

At the top of each KYAML document (or multi-document) will be the following header:

```
#!yaml
---
```

This disambiguates it from JSON, since both KYAML and JSON begin with an opening curly-brace.  Kubernetes versions 1.33 and higher can handle KYAML without the header, but it is needed for versions before that.

### Per-type rendering

Integers and floats are rendered as their natural numeric representation.

Simple strings are rendered as double-quoted and escaped. Multi-line strings require special consideration, see below.

Booleans are rendered as `true` or `false`.

> Structs are rendered with curly-braces. Field names are not quoted unless they are not obviously strings (which includes prefixed label keys) or are one of the known type-ambiguous words (e.g. `no`).  Field values are rendered as per their type. Embedded structs will be rendered as inline fields.  Fields which have a `json` struct tag will be rendered according to that tag:

Reviewer 2: This doesn't say whether a struct will be one long line (`{foo: 1, bar: 2}`) or use one line per field. It becomes clearer later on, but perhaps mention it here so that one can read from top-to-bottom instead of having to jump ahead?

When you touch this line, perhaps reflow it?

Author: ACK

>   * the tag’s field name will be used

Reviewer 2: Doesn't this break encoding a struct as KYAML (uses JSON field name) and then decoding it with a normal YAML decoder (ignores JSON field name)?

I'm okay with this and it makes sense in the context of Kubernetes, but it should be called out somewhere.

Author: Or own YAML decoders all seem happy to decode based on JSON name?  I realize now I have no idea why -- I don't see it in the yaml library.  Do you know @BenTheElder ?

Reviewer: > I think our own YAML decoders work by decoding into any, re-encoding into JSON, then parsing into the struct - or at least something like that.

yeah pretty much.

https://github.com/kubernetes-sigs/yaml#introduction

https://github.com/kubernetes-sigs/yaml/blob/master/yaml.go

Author: So we could make the kyaml encoder respect YAML tags, inasmuch as they exist but ar non-standard, or just say it respects JSON tags because that IS standard.

Reviewer 2: I'm fine with what your proposed (use JSON tags), but let's call out the asymmetry somewhere.

Author: Added a note and an UNRESOLVED to collect opinion

  * if the tag’s name is “-”, the field will be skipped
  * if the tag specifies “omitempty” it will be respected
  * if the tag specifies “omitzero” it will be respected

> Maps are rendered with curly-braces. Keys are sorted and are not quoted unless they are type-ambiguous (e.g. `no`), and values are rendered as per their type.

Reviewer 2: Same question about single-line vs multi-line.

Author: ACK

<<[UNRESOLVED quoted strings]>>
Should we treat strings for field names and map keys:
As described above, with label keys quoted but field names not
Allow unquoted label keys and more complex (but not ambiguous) values
Always quote strings, including ALL map keys and field names
<<[/UNRESOLVED>>

Slices and arrays are rendered with square-braces. Elements are rendered as per this list.

Pointers are rendered as `null` or dereferenced and rendered as per the pointee’s type.

Types which implement the Go standard `json.Marshaler` interface (`MarshalJSON()`) are rendered by marshaling to JSON, unmarshaling into an `any`, and then rendering that `any` as KYAML.

Types which implement the `yaml.Marshaler` interface (`MarshalYAML()`) are rendered as per their type.  `MarshalYAML()` is not a Go standard and is not used by KYAML.

<<[UNRESOLVED yaml self-marshaling]>>
Should we do something for YAML self-marshalling?  It is non-standard and we do not use it in Kubernetes.
<<[/UNRESOLVED>>

Interfaces are rendered as `null` or their true value type.

Any other type (as per `reflect.Value.Kind()`) will produce an error.

### Multi-line strings

> For better or worse, multi-line strings are a thing that people do in Kubernetes, and conventional YAML makes it pretty easy. KYAML has to choose how we want to handle it.  Consider these examples of conventional YAML.

Reviewer 2: This is `UNRESOLVED`, isn't it?

Reflow long line?

Author: ACK

Example 1:
```
example_1: |
    This
    is a
    multi-line string
```

Example 2:
```
example_2: |
    this:
      is:
      - embedded
      - yaml
      where:
        whitespace:
          matters: true
```

This style is not available in YAML flow-style.

#### Option 1. Treat it like a Go quoted-and-escaped string

This always comes out as one line.

```
example_1: “This\nis a\nmulti-line string\n”
example_2: "this:\n  is:\n  - embedded\n  - yaml\n  where:\n    whitespace:\n      matters: true\n"
```

Pro: Easy and well-defined
Pro: Can be pasted into most C-like languages to get the original string.

> Con: Hard to read and write by humans when it gets long.

Reviewer 2: I agree that this is not an option, in particular with an eye towards using KYAML for logging and test output. The original klog structured text log output used this and we promptly got bug reports and reverts because it made the log output less readable. I then added multi-line output.

The problem is that none of the options below seem particularly appealing either :disappointed:

Reviewer: I think 2) is tolerable. All multi-line strings are a bit awkward in yaml, the "flow" style strings have different problems we didn't discuss here because they can't be used with block style.

Reviewer 2: I agree that option 2 is okayish. That `this:` and `  is:` in example 2 are shifted by one character (the `\` before `  is:`) is a pity.

BTW, you seem to use the opposite definition of "flow" and "block" than the KEP? Option 1 is said to be "not available in YAML flow-style" with the string using block style, so options 2 onward must be flow-style strings. Or am I mixing things up? I hadn't seen this terminology before.

Author: flow-style is with `{}` and `[]`

block-style is with whitespace

https://yaml.org/spec/1.2.2/

Section 2.1 says:

"""
YAML also has flow styles, using explicit indicators rather than indentation to denote scope. The flow sequence is written as a comma separated list within square brackets. In a similar manner, the flow mapping uses curly braces.
"""

Section 3.2.3.1 says:

"""
There are two groups of styles. Block styles use indentation to denote structure. In contrast, flow styles rely on explicit indicators.
"""

If we got that wrong anywere, please flag it :)

Reviewer 2: The KEP has it right, but Ben seems to use it the other way around in his comments - or I am misreading them.

Reviewer: [I transposed flow and block in the comment above, I will edit it]

> I agree that option 2 is okayish. That this: and   is: in example 2 are shifted by one character (the \ before   is:) is a pity.

Yeah, though if you're depending on significant leading whitespace *inside of a yaml multi-line string* there's simply no non-error prone format for that particular aspect.

Some of them require less characters, but they're not any less confusing or error prone.

In general if you want to guarantee leading whitespace you need to escape it.

Reviewer 2: > That this: and   is: in example 2 are shifted by one character (the \ before   is:) is a pity.

Variation B of option 2 addresses this nicely - it's my favorite solution.

Reviewer: Agreed.

NO

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

NO

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

Using this may incur some client-side resource usage, but marginal.

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

NO

### Troubleshooting

If someone thinks the KYAML rendering is wrong they can instead render to plain YAML or JSON and compare.  Additionally, they can render to KYAML and then use yamlfmt to convert that to plain YAML.

###### How does this feature react if the API server and/or etcd is unavailable?

N/A

###### What are other known failure modes?

It’s possible we missed some esoteric structure that renders incorrectly.  Those will be handled as bugs.

###### What steps should be taken if SLOs are not being met to determine the problem?

N/A

## Implementation History

May 09, 2025: KEP draft v0

## Drawbacks

Having a local dialect of YAML could be confusing to people who are used to conventional YAML.

> Trying to text-patch KYAML with plain YAML (as in a help chart) will almost certainly not work.

Reviewer 2: ```suggestion
Trying to text-patch KYAML with plain YAML (as in a helm chart) will almost certainly not work.
```

Will not work in which sense? It wouldn't be KYAML anymore, but should still be valid YAML?

Reviewer: If you were writing flow-style YAML and you try to text-substitute it into YAML which has switched to block style it is likely invalid. So if your helm chart decided to switch then you would have to adapt.

That said, text-templating YAML instead of using patches or some other structure aware approach is _always_ fragile.

Author: Added a note

style”, and it corrupts KYAML input today (aside: it also does not handle
multi-documents).  If we intend to convert examples, we need to fix this.

We propose to add flags to it which specify to use “conventional” YAML or KYAML
output, and set the default to KYAML (or to change the parent script to always
specify KYAML).

This tooling will preserve comments.

#### How?

Today we have forked go-yaml.v2 AND go-yaml.v3 into sigs.k8s.io/yaml.  While
KYAML is an output format, our tooling goals require that we get access to a
YAML AST, which may require some dirty tricks.  If we ever change YAML
libraries (e.g.
[https://github.com/goccy/go-yaml](https://github.com/goccy/go-yaml)), we may
need different dirty tricks.

Both sigs.k8s.io/yaml/go-yaml.v2 and v3 offer a way to decode YAML into an AST,
which we can render as KYAML.  Neither, however, offers a way to encode an
arbitrary Go object as the equivalent YAML AST.  To do that we may have to
encode the object to plain YAML, then decode it to AST, then render it as
KYAML.  That extra conversion is ugly, but this is not generally
performance-sensitive code and is not run on the server.

> [https://github.com/goccy/go-yaml](https://github.com/goccy/go-yaml) offers an

Reviewer 3: what is the impact of this dependency? is this going to impact k8s.io/{api, apimachinery or client-go}

Author: It's a significant decision

Author: If we predicate this KEP on adopting goccy I am afraid it will take longer than a single release :)

The amount of fear and defense-in-depth we have established around go-yaml is significant.  @liggitt

API which appears to do what we need in both directions.

### Header

At the top of each KYAML document (or multi-document) will be the following header:

```

> #!yaml

Author: It would have to be an executable file and it would still (probably) fail since it is not actually executable.

This could be anything - the comment isn't even REALLY required, just the `---`

> #!yaml

Reviewer: I hope people are not setting executable bit on their yaml :-)

---
```

This disambiguates it from JSON, since both KYAML and JSON begin with an
opening curly-brace.  Kubernetes versions 1.33 and higher can handle KYAML
without the header, but it is needed for versions before that.

### Per-type rendering

Integers and floats are rendered as their natural numeric representation.

Simple strings are rendered as double-quoted and escaped. Multi-line strings
require special consideration, see below.

Booleans are rendered as `true` or `false`.

Structs are rendered with curly-braces, and each field on a line of its own.
Field names are not quoted unless they are not obviously strings (which
includes prefixed label keys) or are one of the known type-ambiguous words

> (e.g. `no`).  Field values are rendered as per their type. Embedded structs

Reviewer 2: ```suggestion
(e.g. `no`).  Field values are rendered as per their type in the order in which they appear there. Embedded structs
```

This is a nice improvement over our current YAML encoder: because of the intermediate conversion to a map, the original field order gets lost. Not so with KYAML :smile:

Author: Unfortunately, that makes it sound like `kubectl` would handle it properly, but it doesn't.  It is converted to Unstructured before we ever see it (it might be a CRD) and the order of fields is not retained.

Reviewer 2: I'm aware that it won't work for kubectl. I was thinking of the `klog.Dump(<some API type>)` scenario.

Reviewer: It might be confusing to have a different order for the same data depending on the execution context ..

Author: That's what we have today - if you just `json.Marshal` a kube object it will retain field ordering, but if you run it through kubectl it will not.

Reviewer 4: the most reliable way to make sure we perfectly match json typing / omitempty / omitzero / custom marshaler semantics is to use a pipeline like `stdlib json.Marshal(obj) --> yaml order-preserving AST unmarshal → kyaml emitter`

2. Server-side support for KYAML or YAML
3. Requiring projects to migrate by default.

## Proposal

KYAML is a specification of output.
Since KYAML is a subset of YAML,
it should always be accepted as input,
but this KEP does not demand that such input be strict KYAML.
Users can choose to violate this specification,
As long as the result is still valid YAML.

KYAML addresses the problems cited above by:
1. Always quoting _value_ strings (no ambiguity around things like "no").
2. Allowing (but not requiring) quoted keys, but always quoting them if they are not obviously safe (e.g. "no" would always be quoted).
3. Always using {} for structs and maps (no more obscure errors about mapping values).
4. Always using [] for lists (no more trying to figure out if a dash changes the meaning).

Taken together,
these mean that KYAML is not whitespace-sensitive,
which makes templating and other text-patching much easier.
YAML calls this “flow style”,
as opposed to the conventional “block style”.

This might sound a lot like JSON,
because it is!
Remember that JSON is valid YAML.
Unlike JSON, KYAML:

> 1. Allows comments.

Reviewer: This says "unlike JSON", the point is that KYAML does look pretty JSON-ish as a format.
Humans can edit KYAML to add comments and it's still KYAML.
If they use `-o json` as a "less error prone yaml alternative" you can't do these things.

kyaml is output-only _from kubectl_, but it's still effectively an input format as a YAML subset that will make its way into files etc.

> 1. Allows comments.

Author: Good point.  As a format, comments are allowed, so a user can take the result of `kubectl` and add comments and store it, and submit that back to kubectl.  Unlike JSON.  Will clarify

<<[/UNRESOLVED]>>

### Tooling

k/k has a CI verifier (hack/verify-yamlfmt.sh) which looks for YAML files and
ensures they are formatted “correctly”. The only meaning of “correct” is “block
style”, and it corrupts KYAML input today (aside: it also does not handle
multi-documents).  If we intend to convert examples, we need to fix this.

We propose to add flags to it which specify to use “conventional” YAML or KYAML
output, and set the default to KYAML (or to change the parent script to always
specify KYAML).

This tooling will preserve comments.

#### How?

Today we have forked go-yaml.v2 AND go-yaml.v3 into sigs.k8s.io/yaml.  While
KYAML is an output format, our tooling goals require that we get access to a
YAML AST, which may require some dirty tricks.  If we ever change YAML
libraries (e.g.
[https://github.com/goccy/go-yaml](https://github.com/goccy/go-yaml)), we may
need different dirty tricks.

Both sigs.k8s.io/yaml/go-yaml.v2 and v3 offer a way to decode YAML into an AST,
which we can render as KYAML.  Neither, however, offers a way to encode an
arbitrary Go object as the equivalent YAML AST.  To do that we may have to

> encode the object to plain YAML, then decode it to AST, then render it as

Reviewer 4: I would expect us to encode to json using the standard json encoder, then decode, then transform / render as kyaml. That gets us all the json encoding behaviors for ~free.

Author: It also loses comments.  If you run `yamlfmt --kyaml` (or whatever we end up with) on a file, it should reflow it to KYAML but retain the comments.

Reviewer 4: I think there's two entry points to the pipeline:
1. In-memory object (the PrintObject entry point).
2. Yaml bytes (which includes json bytes) we want to transform to kyaml (from files or arbitrary inputs)

If we handle entry point 1 by serializing to JSON, we'll perfectly match all JSON semantics and behavior (and will preserve struct order for typed inputs and sort keys for unstructured / map inputs, just like -o json), and can then route that output to entry point 2, reflowing it into kyaml format.

Author: Maybe I am missing something.

Case 1: From a Go object.
- Call `json.Marshal()` on it (preserves field order).
- Call `json.Unmarshal()` to decode to a map (loses field order).
- Call `kyaml.Marshal()` 

?  Am I misunderstanding?  Or maybe you are assuming that we can convert JSON to an AST?  I guess I could try through `json.Decoder.Token()` and re-encode that -- is that what you meant?

Case 2a: From JSON bytes.
- Use Decoder as suggested above

Case 2b: From YAML bytes.
- Decode to `yaml.Node` (retains comments)
- Render to kyaml.

That leaves me with 2 paths still?

Reviewer 4: I was envisioning:

Case 1: From a Go object.

Call json.Marshal() on it (preserves field order). The output should be passable to case 2, preserving order.

Case 2: From JSON or YAML bytes.

Decode to yaml.Node (retains comments)
Render to kyaml.

Author: Oh, or maybe you meant:

Case 1: From a Go object.
- Call `json.Marshal()` on it (preserves field order).
- Call `yaml.Unmarshal(*yaml.Node)` on the JSON bytes
- Render Node -> KYAML

Case 2: From JSON bytes.
- Call `yaml.Unmarshal(*yaml.Node)` on the JSON bytes
- Render Node -> KYAML

Case 3: From YAML bytes.
- Call `yaml.Unmarshal(*yaml.Node)` on the YAML bytes (preserve comments)
- Render Node -> KYAML

That could work.

Possible problem: JSON "streams" are not supported by YAML as multi-documents.  That's PROBABLY fine?

Author: Mid-air collision.  That may be possible.  I had "forgotten" that i can pass JSON bytes to yaml library (unless it is a multi-document stream)

Reviewer 2: Another advantage of first invoking the JSON encoder: it protects against types and data structures which cannot be encoded (keys that are not supported, cycles). Otherwise that has to be re-implemented by the KYAML encoder.

Author: Detailed in the next KEP push

without the header, but it is needed for versions before that.

### Per-type rendering

Integers and floats are rendered as their natural numeric representation.

Simple strings are rendered as double-quoted and escaped. Multi-line strings
require special consideration, see below.

Booleans are rendered as `true` or `false`.

Structs are rendered with curly-braces, and each field on a line of its own.
Field names are not quoted unless they are not obviously strings (which
includes prefixed label keys) or are one of the known type-ambiguous words
(e.g. `no`).  Field values are rendered as per their type.  Embedded structs
will be rendered as inline fields.

KYAML does not guarantee that fields will be rendered in the order they are
defined in the struct.  Implementations may preserve field order, but `kubectl`
itself does not guarantee this (it converts to a map internally, which loses
the type schema), so the order may be different.

Fields which have a `json` struct tag will be rendered according to that tag:
  * the tag’s field name will be used
  * if the tag’s name is “-”, the field will be skipped
  * if the tag specifies “omitempty” it will be respected

>   * if the tag specifies “omitzero” it will be respected

Reviewer 4: Ordering (of order-independent bits) and formatting can differ, but the content and structure of the output must be identical to `-o json` and `-o yaml`, right?

Rendering as json using the standard json encoder and then transforming into kyaml seems like the most bullet-proof way to accomplish this

Author: kubectl turns everything into `unstructured` long before the final `Printer` sees the data.  Unstructured implements `json.Marshaler`, which my POC supports directly, so everything gets bounced thru JSON.

I suppose a different approach would be to just bounce EVERYTHING thru JSON, which means I don't have to implement the tag logic (already done, but still).  It means we ALWAYS pessimize the field ordering, no matter where you apply this code.  And it would always be as slow as possible.

I could go either way.  It's EFFECTIVELY what is happening already in kubectl.

Reviewer 4: I think the generic kubectl commands all use unstructured, but some of the commands do use typed objects as output. The more we can do to auto-inherit JSON serialization semantics and make this *just* do yaml reflowing, the better

Reviewer 2: > kubectl turns everything into `unstructured` long before the final Printer sees the data. 

Random thought (unrelated to this KEP): what if kubectl used something else than `unstructured` which implements the same interfaces (for example, meta data access) but preserves field ordering? I'm tempted to try it out...

Reviewer 2: Probably, but it would be a major breaking change of our API. That it's a map is exposed: https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured#Unstructured

I suppose someone noticed that this might not be a good idea:
> TODO: make the serialization part of this type distinct from the field accessors.

Reviewer: You can do some really surprising things in YAML map keys that we should be aware of but may choose to ignore here. YAML only requires that keys be unique. Ideally we want KYAML to be fairly reasonable beyond just kubectl printing though as we get the most benefit when it's generally usable in the ecosystem to prefer formatting this way.

we somewhat address this already with the "sometimes quote keys"

The `yaml` struct tag will be ignored.

<<[UNRESOLVED yaml self-marshaling]>>

> Should we respect `yaml` struct tags?  It is non-standard and we do

Author: We don't use `yaml` tags in our APIs because we convert it all to JSON along the way.  If the goal is to provide a generally useful KYAML encoder, does that mean we MUST bounce thru JSON?  kubectly is going to do what kubectl does, regardless.

Does not supporting `yaml` tags (even just for the name?) damage our utility?  Probably not.  I am happy to call this resolved if nobody objects.

If I were writing a YAML library I would accept `json` tags and respect them.  I am NOT writing a full YAML library.

not use it in Kubernetes.  Maybe just the name part of it?  Or the
JSON-identical parts?
<<[/UNRESOLVED>>

Maps are rendered with curly-braces, and each key on a line of its own. Keys
are sorted and are not quoted unless they are type-ambiguous (e.g. `no`), and

> values are rendered as per their type.

Reviewer 4: we only support string keys and field names, right?

Author: in kube API we only have strings.  My POC code does not currently handle struct keys, but it could.

Author: Easy to add - they now render as "compact" forms of their natural type, e.g.:

```
+               // weirdly-keyed maps
+               {"struct-key map", map[kv]int{
+                       {"k1", "v1"}: 1,
+                       {"k2", "v2"}: 2,
+               }, "{\n  { K: \"k1\", V: \"v1\" }: 1,\n  { K: \"k2\", V: \"v2\" }: 2,\n}"},
+               {"list-key map", map[[2]string]int{
+                       {"s1", "s2"}: 12,
+                       {"s3", "s4"}: 34,
+               }, "{\n  [ \"s1\", \"s2\", ]: 12,\n  [ \"s3\", \"s4\", ]: 34,\n}"},
+               {"multi-line-string-key map", map[string]int{
+                       "1\n 2\n  \n3": 123,
+                       "4\n 5\n  \n6": 456,
+               }, "{\n  \"1\\n 2\\n  \\n3\": 123,\n  \"4\\n 5\\n  \\n6\": 456,\n}"},
```

Reviewer 4: No non-string key will ever work in a kube REST or config API. I wouldn't try to support them.

If we route through JSON to create output from an in-memory object, it'll never be an issue.

If we catch/error on non-string keys when reflowing existing yaml input to kyaml, that seems ideal

Author: If we define KYAML as the strictly JSON-compatible subset of YAML features, we can get away with this.

> <<[UNRESOLVED quoted strings]>>

Reviewer 4: categories of known problems for unquoted keys / field names:
* things that look like booleans
* things that look like integers / floats / scientific notation
* things that look like structure (whitespace, newlines, internal colons, special characters, etc)
* things that trigger weird yaml behaviors (`!` directives, boolean casting, etc)
* others...?

possible places to land on the spectrum:
* "Always quote": by far the simplest and easiest to verify correctness of.
* "Leave simple verifiably safe things unquoted, err on the side of quoting everything else, possibly unnecessarily": less simple, still mostly readable, plausible to verify correctness of
* "Only quote when absolutely necessary": most readable, I guess?

Author: Today, my POC code does the 2nd one: "Leave simple verifiably safe things unquoted, err on the side of quoting everything else".  My "verifiably safe" logic is:

```
func yamlGuessesWrong(s string) bool {    
    // Null-like strings: https://yaml.org/type/null.html    
    switch s {    
    case "null", "Null", "NULL", "~", "":    
        return true    
    }    
    
    // Boolean-like strings: https://yaml.org/type/bool.html    
    if _, err := strconv.ParseBool(s); err == nil {    
        return true    
    }    
    switch strings.ToLower(s) {    
    case "true", "y", "yes", "on", "false", "n", "no", "off":    
        return true    
    }    
    
    // Number-like strings: https://yaml.org/type/int.html and    
    // https://yaml.org/type/float.html    
    //    
    // NOTE: the stripping of underscores is gross.    
    sWithoutUnderscores := strings.ReplaceAll(s, "_", "")    
    // Handles binary ("0b"), octal ("0" or "0o"), decimal, and hex ("0x")    
    if _, err := strconv.ParseInt(sWithoutUnderscores, 0, 64); err == nil && !isSyntaxError(err) {    
        return true    
    }    
    // Handles standard and scientific notation.    
    if _, err := strconv.ParseFloat(sWithoutUnderscores, 64); err == nil && !isSyntaxError(err) {    
        return true    
    }    
    
    // Sexagesimal strings like "11:00" (in YAML 1.1, removed in 1.2):    
    // https://yaml.org/type/int.html and https://yaml.org/type/float.html    
    if sexagesimalRE.MatchString(s) {    
        return true    
    }    
    
    // Infinity and NaN: https://yaml.org/type/float.html    
    switch strings.ToLower(s) {    
    case ".inf", "-.inf", "+.inf", ".nan":    
        return true    
    }    
    
    // Time-like strings    
    if _, matches := parseTimestamp(s); matches {    
        return true    
    }    
    
    return false                                                                                                                                                       
}    
```
Timestamps are not completely handled yet.

I might extend it to include label and annotation keys (allowing "." and "/" as long as it doesn't parse as a number).

Should we treat strings for field names and map keys:
As described above, with label keys quoted but field names not
Allow unquoted label keys and more complex (but not ambiguous) values
Always quote strings, including ALL map keys and field names
<<[/UNRESOLVED>>

Slices and arrays are rendered with square-braces. Elements are rendered as per
this list.

Pointers are rendered as `null` or dereferenced and rendered as per the
pointee’s type.

Types which implement the Go standard `json.Marshaler` interface
(`MarshalJSON()`) are rendered by marshaling to JSON, unmarshaling into an
`any`, and then rendering that `any` as KYAML.

Types which implement the `yaml.Marshaler` interface (`MarshalYAML()`) are
rendered as per their type.  `MarshalYAML()` is not a Go standard and is not
used by KYAML.

<<[UNRESOLVED yaml self-marshaling]>>
Should we do something for YAML self-marshalling?  It is non-standard and we do

> not use it in Kubernetes.

Reviewer 4: I don't think we should diverge from `-o yaml`

Author: Any objections from others?

        image: "k8s.gcr.io/serve_hostname:v1.4",
        imageID: "sha256:b56f22170982bc46dee99f7157be69dce727d3b89ee5698d15fda7a14a124200",
        lastState: {},
        name: "serve-hostname-xqnl5",
        ready: true,
        resources: {
          requests: {
            cpu: "100m",
          },
        },
        restartCount: 0,
        started: true,
        state: {
          running: {
            startedAt: "2025-05-09T21:14:41Z",
          },
        },
        volumeMounts: [{
          mountPath: "/var/run/secrets/kubernetes.io/serviceaccount",
          name: "kube-api-access-g57rz",
          readOnly: true,
          recursiveReadOnly: "Disabled",
        }],
      }],
      hostIP: "10.40.0.9",
      hostIPs: [{
        ip: "10.40.0.9",
      }],
      phase: "Running",
      podIP: "10.64.3.4",
      podIPs: [{
        ip: "10.64.3.4",
      }],
      qosClass: "Burstable",
      startTime: "2025-05-09T21:14:39Z",
    },
  }],
  kind: "List",
  metadata: {
    resourceVersion: "",
  },
}
```

### Test Plan

The new encoder will include robust tests to prove both its own output format

> and its compatibility with standard YAML.

Reviewer 4: and consistency with `-o json`

1. Always quoting _value_ strings (no ambiguity around things like "no").
2. Allowing (but not requiring) quoted keys, but always quoting them if they are not obviously safe (e.g. "no" would always be quoted).
3. Always using {} for structs and maps (no more obscure errors about mapping values).
4. Always using [] for lists (no more trying to figure out if a dash changes the meaning).

Taken together,
these mean that KYAML is not whitespace-sensitive,
which makes templating and other text-patching much easier.
YAML calls this “flow style”,
as opposed to the conventional “block style”.

This might sound a lot like JSON,
because it is!
Remember that JSON is valid YAML.
KYAML is friendlier than JSON for humans
to read,
to write,
and to commit to source control.
Unlike JSON, KYAML:
1. Allows comments (when authored by users).
2. Allows trailing commas on the last item in a list or struct/map.
3. Does not require quoted keys.

KYAML also includes a “header” (still valid YAML)
which helps to disambiguate a KYAML document from an ill-formed JSON document
(see: [kubernetes/kubernetes#77946](https://github.com/kubernetes/kubernetes/issues/78946)).

> KYAML also tries to economize on vertical space by ["cuddling"](https://blog.gskinner.com/archives/2008/11/curly_braces_to.html) some kinds of brackets together.

Reviewer 4: For the reflowing of existing yaml with comments, need to think through how comments attached at various points in the input will be moved/combined and where they will end up in the output

Author: I think it's OK.  A yaml.Node has attach points before, same-line, and after.  Assuming what you meant is to transmute a yaml.Node.  If you have some other idea, I am still missing it.

Reviewer 4: I'm sure it's tractable, just trying to envision / pin down where comments like these:
```
# list key, before
items: # list key, same line
# item 1, before
- key1: "1" # key1, same line
  # key2, before
  key2: "2" # key2, same line
  # key2, after
# item 2, before
- key1: "1" # key1, same line
# item 2, after
```

when reflowed to:

```
items: [{
  key1: "1",
  key2: "2",
}, {
  key1: "1",
  key2: "2",
}]
```

Especially the comments between items that can't fit between the "}, {" bits, etc.

I'm pretty sure comments can be consolidated / placed reasonably, but a test case with inputs hanging comments every place we can imagine showing how we preserve or combine or relocate them would be a good idea.

Author: Yeah, I had not considered comments in the specifications because the yamlfmt topic was added after I wrote the bulk.  Let me see if I can hack up a pic of the node rendering.  Comments would change some of the cuddling rules, probably

be rendered as inline fields.

KYAML does not guarantee that fields will be rendered in the order they are
defined in the struct.  Implementations may preserve field order, but `kubectl`
itself does not guarantee this.  For example, `kubectl` converts everything to an
unstructured type internally, which loses the field ordering.

Fields which have a `json` struct tag will be rendered according to that tag,
by nature of the fact that we marshal Go objects to JSON first.

The `yaml` struct tag will be ignored.

##### Maps

Maps are rendered with curly-braces, and each key on a line of its own. If a
map has a line-comment it will be rendered after the closing brace (since
this is the only place it will be found upon re-reading).

Because we marshal Go objects to JSON, map keys must be strings. This is more
restrictive than plain YAML, meaning that reformatting YAML with non-string
keys into KYAML will fail.

Keys are sorted and are not quoted unless they are not obviously strings or are
one of the known type-ambiguous words, the same as for struct fields.

Values are rendered as per their type.

##### Lists

Slices and arrays are rendered with square-braces. If a struct has a

> line-comment it will be rendered after the opening brace.

Author: Yeah, it was retrfitted and I didn't feel like it warranted a whole new section - will make it clearer

List elements are rendered as per their type.

##### Pointers

Pointers are rendered as `null` or dereferenced and rendered as per the
pointee’s type.

##### Self marshalers

Types which implement the Go standard `json.Marshaler` interface
(`MarshalJSON()`) are rendered by marshaling to JSON.

Types which implement the `yaml.Marshaler` interface (`MarshalYAML()`) are
rendered as per their type.  `MarshalYAML()` is not a Go standard and is not
used by KYAML.

##### Interfaces

Interfaces are rendered as `null` or their true value type.

##### Anything else

Any other type (as per `reflect.Value.Kind()`) will produce an error.

#### Multi-line strings

For better or worse, multi-line strings are a thing that people do in
Kubernetes, and conventional YAML makes it pretty easy. KYAML has to choose how
we want to handle it.  Consider these examples of conventional YAML.

Example 1:
```
example_1: |
    This
    is a
    multi-line string

> ```

Author: I think the example is correct?

```
$ cat /tmp/ex.yaml 
example_1: |
    This
    is a
    multi-line string

$ yq < /tmp/ex.yaml 
{
  "example_1": "This\nis a\nmulti-line string\n"
}
```

Example 2:
```
example_2: |
    this:
      is:
      - embedded
      - yaml
      where:
        whitespace:
          matters: true
```

This style is not available in YAML flow-style.

KYAML will use YAML flow-folding and escaped line-breaks to get "close" to the
original without sacrificing the main goals of this KEP.  A flow-fold is a
trailing backslash at the very end of a line, which tells YAML to remove the
newline and any leading whitespace on the next line.

To get newlines we will use `\n` in the string.

To get leading whitespace, we will use an escape character, which tells YAML
that any further whitespace is part of the contents.

We will add an extra space on lines which do not have leading whitespace, to
preserve vertical alignment.

Note: the extra space is only in the rendering, not in the actual content,
since YAML drops all leading whitespace until the first non-whitespace
character.

Example:

```
example_1: ”\
     This\n\
     is a\n\
     multi-line string\

>     “

Author: Ugh, thank you.

example_2: "\
     this:\n\
    \  is:\n\
    \  - embedded\n\
    \  - yaml\n\
    \  where:\n\
    \    whitespace:\n\
    \      matters: true\n\
    "
```

### Commas

Lists, maps, and structs always have a trailing comma after their final
element, unless the closing brace is cuddled.  As mentioned elsewhere, the
comma is not required, but KYAML always emits it.

Example:

```
labels: {
  app: "foobar",
  foo: "bar",
  something: “12345”,
},
```

### Braces

Paired braces are always cuddled, unless the next item has a head-comment or
the previous item has a foot-comment.

This is best illustrated with an example:

Cuddled:
```
list: [{
  foo: “abc”, # This is a line-comment
  bar: 123,
}, {
  foo: “def”,
  bar: 456,
}]
```

Partially cuddled, with a head-comment.
```
list: [
# This is a head-comment
{
  foo: “abc”,
  bar: 123,
}, {
  foo: “def”,
  bar: 456,
}]
```

Partially cuddled, with a foot-comment.
```
list: [{
  foo: “abc”,
  bar: 123,
},
# This is a foot comment
{
  foo: “def”,
  bar: 456,
}]

> ```

Author: go-yaml has weird options for comments.  We just have to live with them for now.  Will clarify.

And yes head-vs-foot is ambiguous when there are no blank lines.

### Indents

Indents follow YAML’s two-space convention.

### Short lists, maps, and structs

No affordance is made to render short lists, maps, or structs on a single line.

> This is possible but was deemed unnecessary.

Reviewer 5: (I assume *empty* lists/maps/structs are still rendered on a single line?)

decoder), and compares the original and result objects. Doing this enough times
should give us fairly high confidence.

##### Integration tests

N/A

##### e2e tests

N/A

### Graduation Criteria

This KEP introduces a new env-var feature gate to prevent premature
dependencies from forming.

#### Alpha

The new env-var feature gate will be checked before allowing use of the "kyaml"
output option. If `$KUBECTL_KYAML` is not set to "true", the "kyaml" output
option will not be available.

- The `kubectl` output format `kyaml` is implemented
- The gate is disabled by default
- All unit tests are complete
- Gather feedback from developers and surveys

#### Beta

The new env-var feature gate is enabled by default, but still disableable. If
`$KUBECTL_KYAML` is not set to "false", the "kyaml" output option will be
available.

- Resolve any identified issues.

#### GA

The new env-var feature gate is removed, the "kyaml" output option is always
available.

- A proper specification is published
- The majority of examples in the kubernetes GitHub orgs are converted to KYAML
- No new bug reports

This KEP does not propose to add conformance tests.

### Upgrade / Downgrade Strategy

> This is a client-side feature.  Users should not use this in automation until

Reviewer 5: ```suggestion
This is a client-side feature.  Users should not use `-o kyaml` in automation until
```

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

NO

### Troubleshooting

If someone thinks the KYAML rendering is wrong they can instead render to plain
YAML or JSON and compare.  Additionally, they can render to KYAML and then use
yamlfmt to convert that to plain YAML.

###### How does this feature react if the API server and/or etcd is unavailable?

N/A

###### What are other known failure modes?

It’s possible we missed some esoteric structure that renders incorrectly.
Those will be handled as bugs.

###### What steps should be taken if SLOs are not being met to determine the problem?

N/A

## Implementation History

May 09, 2025: KEP draft v0
May 26, 2025: KEP draft implementable

## Drawbacks

Having a local dialect of YAML could be confusing to people who are used to
conventional YAML.

Trying to text-patch KYAML with plain YAML (as in a helm chart) will almost
certainly not work (mixing flow- and block-styles is a mess).

More serialization “languages” dilutes knowledge.

[Yet another standard](https://xkcd.com/927/). Rebuttal: KYAML is YAML and

> users can provide vanilla YAML anyywhere that can accept KYAML.

Reviewer 5: ```suggestion
users can provide vanilla YAML anywhere that can accept KYAML.
```

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [X] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP proposes to add a new output format for kubectl,
called "KYAML"
(as in `kubectl get -o kayml ...`).
This format is a strict subset (aka "dialect") of standard YAML,
and so should be parseable by the existing ecosystem.
This dialect seeks to emphasize syntactical choices which avoid
many of the most common traps in YAML.
For example, unlike standard YAML output,
this dialect is not whitespace-sensitive,

> which makes it vastly easier to patch correctly in things like Helm charts.

Reviewer 6: I brought up this KEP in the Helm Weekly Developer Meeting which I regularly participate.
I'd like to get Helm release manage @mattfarina 's take on all of it noted here.

My take is that while you are saying this is optional, you really are suggesting that people use/recommend/document an entirely new format, that while technically valid YAML, will not be recognized by most users as such.
YAML has been in k8s since the start. That's 10+ years of establishment that is trying to be changed.
And for what I see is: trading a couple common issues for a whole new set of them. 

cc: @banjoh @scottrigby @gjenkins8 from https://github.com/helm/helm

Author: I'm fine to take feedback, but I don't think we should block merge on that.  The KEP window closes soon, and I am travelling in a week, so I don't want to miss it on technicality.

Reviewer 6: When does the KEP window close?

This KEP further proposes to make KYAML the standard format

> for all project-owned documentation and examples.

Reviewer 6: As the YAML data language project and the yaml/go-yaml project lead, I want to state clearly that:
**I do not endorse the KYAML proposed format at this time**.

While I am interested in exploring the topic and contributing my comments, opinions and tests, I don't currently feel like this is a good direction overall.

It fragments YAML best practices, by using specific features of the language in ways that are not at all common and that give up some of the things that make YAML popular, in search of solutions for problems that new users of YAML can fall prey to. 

The new go-yaml project is looking to bring YAML initiatives in Go closer together. Less forked. Address YAML issues across the board, rather than come up with more YAML snowflake solutions.

I would much rather see (human maintained) YAML made files become more composable and functional, and have a lot of ideas to offer there.

I don't have a problem with programs _emitting_ YAML in defined/configurable styles (like KYAML), but I think this is best served by a configurable `yamlfmt` CLI and the same configurability available for go-yaml's Marshall. This would effectively turn "KYAML" into a go-yaml Marshall configuration. 

But I really can't get behind encouraging people to write/maintain YAML this way.

Reviewer 6: I also want to add that I think there are a lot of YAML cases that haven't been addressed here yet.

I've been traveling for much of the time I've taken to contribute here, but haven't had the chance to really dig in.

I'd be happy to offer more next week when I'm settled.

## Motivation

The vast majority of examples of how to use Kubernetes include YAML.
These all use the most "conventional" YAML syntax,
which is what `kubectl get -o yaml` also uses.
Simple YAML is easy to read, but complex YAML can become very tricky.
Some of the things that motivated this KEP are:

Whitespace being significant means that
readers and writers must track the nesting depth.
While this does not sound too onerous,
We have ample evidence that it is, in fact, very hard.
Wrongly indented YAML is syntactically valid but does not marshal properly.
Perhaps worst of all,
template-expanding systems,
such as the widely used `helm`,

> force users to get whitespace "just right"

Reviewer 6: Helm charts done with kyaml would be much less readable/maintainable with all the curlies already required by Go templating.

Also while indentation isn't a parsing issue in kyaml, it certainly is an issue wrt readability.
It's disingenuous to say that the `nindent`s will go away.
They won't since the output of `helm template ...` needs to be human readable/parsable.

Author: helm template | yamlfmt (or yq) works

Reviewer 7: I would argue the development of Helm templates will be more difficult. Most users of Helm never look at the generated output. They just see what's installed.

My concern would be that YAML structures and practices for k8s (helm included) are already widely set. Making changes to that is going to cause confusion.

The developer experience for Helm templates, and k8s yaml based things in general, will go down by using a style that's not widely used or documented in the world at large.

and reduce common mistakes and frustration.

### Goals

1. Specify a YAML dialect which is 100% compatible with existing parsers and tooling.
2. Provide libraries and tooling for encoding and converting to this formatting, with
   stable (idempotent) conversion.
3. Provide a KYAML output format for kubectl.
4. Update examples and documentation throughout the project.
5. Address common YAML pitfalls through an opinionated formatting approach.
6. Document the design and reasons for KYAML and the 100% compatibility with tooling expecting arbitrary YAML.

### Non-Goals

1. Introduce alternative configuration languages that are not compatible with existing tooling.
2. Server-side support for KYAML distinct from what we support in YAML already.
3. Require projects to migrate to KYAML by default.
4. Change the YAML library used by Kubernetes.

## Proposal

KYAML is primarily a specification of output.
Since KYAML is a subset of YAML,
it should always be accepted as input,
but this KEP does not demand that such input be strict KYAML.
Users can choose to violate this specification,
As long as the result is still valid YAML.

KYAML addresses the problems cited above by:

> 1. Always quoting _value_ strings (no ambiguity around things like "NO").

Reviewer 6: This document should say "double-quoting" instead of quoting, to be more accurate.
There are multiple ways to quote scalars in YAML, each with different semantics.

Author: ACK

which helps to disambiguate a KYAML document from an ill-formed JSON document
(see: [kubernetes/kubernetes#78946](https://github.com/kubernetes/kubernetes/issues/78946)).

KYAML also tries to economize on vertical space by ["cuddling"](https://blog.gskinner.com/archives/2008/11/curly_braces_to.html) some kinds of brackets together.

Because KYAML is YAML, a KYAML multi-document is a YAML multi-document.

To make this change, we will:
1. Create new code (see details below) which serializes any object into KYAML.
2. Introduce a kubectl output format, `-o=kyaml`, which uses this new package.
3. Create tooling which reformats YAML or JSON as KYAML (see details below).
4. Work with SIG Docs to update examples to this format, both example command-lines and examples of API objects.
5. Look for other examples across kubernetes-affiliated git repositories, and offer PRs to convert to KYAML.
6. Make a lot of noise about it, and try to get more people to use it

The rationale for changing everything everywhere is simple.
Many users start with examples,
which they copy and modify.
By changing the examples we seed the change in the ecosystem.
One success metric might be

> that KYAML becomes the preferred format in the most common helm charts.

Reviewer 6: I don't see how adopting KYAML for Helm chart templates is a tenable idea.
 
> example https://github.com/bitnami/charts/blob/main/bitnami/appsmith/templates/client/deployment.yaml

Getting the Go templating to work correctly is already challenging.
Adding more curly braces to the curly brace soup is not going to be a nice DX.
Cuddling closing YAML curlies makes it worse.

Author: I adapted some of that helm chart and it's (IMO) easier to read and gets rid of all the `| nindent 8` crap.

Reviewer 6: That is about all it can get rid of while adding a lot more curlies.

But I would guess that if kyaml is used with Go templating people will still want the indentation to make `helm template ...` output readable, right?

that we get access to a YAML AST.  Fortunately, go-yaml.v3 offers a way to
decode YAML into such an AST. If we ever change YAML libraries (e.g.
[https://github.com/goccy/go-yaml](https://github.com/goccy/go-yaml)), we will
need to rewrite this code.

### Comments

When reformatting existing YAML with comments, we will attempt to place comments
in "obvious" places.  For example, a line-comment on a scalar field is
obviously placed at the end of the line.  These cases will not be called out in
the subsequent sections of this KEP.  In some cases, the comment placement
may not be obvious, and so will be called out explicitly.

### Disambiguation from JSON

At the top of each KYAML document (or multi-document) we will put a document
separator (`---`) "header". This disambiguates it from JSON, since both KYAML
and JSON begin with an opening curly bracket.  Kubernetes versions 1.33 and
higher can handle KYAML without this, but it is needed for versions before
that.

### Per-type rendering

#### Scalars

Integers and floats are rendered as their natural numeric representation.

Simple strings are rendered as double-quoted and escaped. Multi-line strings

> require special consideration, see below.

Reviewer 6: Double-quoting is often problematic in YAML.

Using them for all string values will require backslash escaping of all double-quotes and backslashes, so code snippets will get harder to write, read and understand.

Given the backslash requirements that KYAML adds for multiline scalars, it will get messier.

Imagine a KYAML formatted scalar that itself is a KYAML formatted multiline scalar containing code with double-quotes and backslashes. 
The backlashes will become problematic. Especially for newer users for which KYAML was created to protect.

YAML's literal scalar syntax allows you to encode any multiline text string (including other YAML files) by simply indenting.

Allowing single-quoting could help a little here. YAML's single quote only needs escaping for one character; the single quote.
Multilines wouldn't work the same way with single-quotes, as double-quotes.

Author: > YAML's literal scalar syntax allows you to encode any multiline text string (including other YAML files) by simply indenting.

Not once you are inside a flow, though?

```
go run ./cmd/yamlfmt/
foo: {
  bar: "quoted",
  qux: |
    non-quoted
failed to decode input: yaml: line 3: found character that cannot start any token
exit status 1
```

Reviewer 6: Right, I was just pointing out that enforcing an all flow style, forgoes literal, which is likely going to be an ongoing pain point for this formatting.

`NO`, `no`, `N`, `YES`, `yes`, `Y`, `On`, and `Off` are all treated as bool
(aka  ["The Norway Bug"](https://hitchdev.com/strictyaml/why/implicit-typing-removed/)),
`_42`, `_4_2_`, and `__________4__________2__________` are all treated as numbers,
and `11:00` is also treated as a number (base 60).

As an alternative to YAML most of Kubernetes also supports JSON,
But JSON has its own drawbacks.
It does not support comments.
It is very particular about trailing commas.
It requires quoted keys.

In addition to `kubectl` output,
most Kubernetes config files are either YAML or JSON.
JSON lacks comments,
which are very valuable for configuration,
so most configuration and tooling is built around YAML.

The most frustrating part of YAML is that it doesn’t have to be this bad.
YAML is a very broad specification.
Inside that specification is a reasonable grammar screaming to be let out.

If we change our tooling and examples to use a less error-prone dialect of YAML,
we can encourage the ecosystem to avoid some of the YAML pitfalls
and reduce common mistakes and frustration.

### Goals

1. Specify a YAML dialect which is 100% compatible with existing parsers and tooling.

> 2. Provide libraries and tooling for encoding and converting to this formatting.

Author: sure

> 2. Provide libraries and tooling for encoding and converting to this formatting.

Reviewer 6: What is the URL for this `yamlfmt`?

> 2. Provide libraries and tooling for encoding and converting to this formatting.

Author: In my working branch:  https://github.com/thockin/kubernetes/tree/kyaml/cmd/yamlfmt

3. Always using {} for structs and maps (no more obscure errors about mapping values).
4. Always using [] for lists (no more trying to figure out if a dash changes the meaning).

Taken together,
these mean that KYAML is not whitespace-sensitive,
which makes templating and other text-patching much easier.
YAML calls this "flow style",
as opposed to the conventional "block style".

This might sound a lot like JSON,
because it is!
Remember that JSON is valid YAML.
KYAML is friendlier than JSON for humans
to read,
to write,
and to commit to source control.
Unlike JSON, KYAML:
1. Allows comments (when authored by users).
2. Allows trailing commas on the last item in a list or struct/map.
3. Does not require quoted keys.

KYAML also includes a "header" (still valid YAML)
which helps to disambiguate a KYAML document from an ill-formed JSON document
(see: [kubernetes/kubernetes#78946](https://github.com/kubernetes/kubernetes/issues/78946)).

KYAML also tries to economize on vertical space by ["cuddling"](https://blog.gskinner.com/archives/2008/11/curly_braces_to.html) some kinds of brackets together.

Because KYAML is YAML, a KYAML multi-document is a YAML multi-document.

> To make this change, we will:

Reviewer 8: hah, I was wondering if this would be the answer. Works for me!

1. Create new code (see details below) which serializes any object into KYAML.
2. Introduce a kubectl output format, `-o=kyaml`, which uses this new package.
3. Create tooling which reformats YAML or JSON as KYAML (see details below).
4. Work with SIG Docs to update examples to this format, both example command-lines and examples of API objects.
5. Look for other examples across kubernetes-affiliated git repositories, and offer PRs to convert to KYAML.
6. Make a lot of noise about it, and try to get more people to use it

> 

Author: encourage, sure, require: no?  Does this KEP need to say that?

defined in the struct.  Implementations may preserve field order, but `kubectl`
itself does not guarantee this.  For example, `kubectl` converts everything to an
unstructured type internally, which loses the field ordering.

Fields which have a `json` struct tag will be rendered according to that tag,
by nature of the fact that we marshal Go objects to JSON first.

The `yaml` struct tag will be ignored (as with our current yaml rendering).

##### Maps

Maps are rendered with curly brackets, and each key on a line of its own.

When reformatting existing YAML, a map's line-comment will be rendered after
the closing bracket (since this is the only place it will be found by go-yaml
upon re-reading).

Because we marshal Go objects to JSON, map keys must be strings. This is more
restrictive than plain YAML, meaning that reformatting YAML with non-string
keys into KYAML will fail.

Keys are sorted and are not quoted unless they are not obviously strings or are
one of the known type-ambiguous words, the same as for struct fields.

Values are rendered as per their type.

##### Lists

Slices and arrays are rendered with square brackets.

> When reformatting existing YAML, a map's line-comment will be rendered after

Reviewer 9: This still holds.

# head1
{
  "f": 1,
},
# head2
{
  "f": 2,
}, {
  "f": 3,
}],
```

Partially cuddled, with line-comments.
```
with_line_comments: [{
  "f": 1,
}, # line1
{
  "f": 2,
}, # line2
{
  "f": 3,
}],
```

Partially cuddled, with foot-comments.
```
with_foot_comments: [{
  "f": 1,
},
# foot1
{
  "f": 2,
},
# foot1
{
  "f": 3,
}],
```

Note that without whitespace a "foot comment" is ambiguous to the the item's
"head comment".  That is an internal issue for go-yaml and should not affect
the rendering, unless we choose to emit blank lines.

### Indents

Indents follow YAML’s two-space convention.

### Short lists, maps, and structs

Empty lists and maps are rendered as `[]` and `{}` respectively.

Other than the special-case above, no affordance is made to render short lists,
maps, or structs on a single line. This is possible but was deemed unnecessary.

### Alignment

No affordance is made to align the values of struct fields or map keys (as gofmt does).

### Consistency of quoted keys

No affordance is made to ensure that keys of a map or struct are all quoted or
not-quoted similarly. Some keys can be quoted while others are not.

### Examples

```
$ kubectl get -o kyaml svc hostnames

> # YAML

Reviewer 9: This still holds.

        volumeMounts: [{
          mountPath: "/var/run/secrets/kubernetes.io/serviceaccount",
          name: "kube-api-access-g57rz",
          readOnly: true,
          recursiveReadOnly: "Disabled",
        }],
      }],
      hostIP: "10.40.0.9",
      hostIPs: [{
        ip: "10.40.0.9",
      }],
      phase: "Running",
      podIP: "10.64.3.4",
      podIPs: [{
        ip: "10.64.3.4",
      }],
      qosClass: "Burstable",
      startTime: "2025-05-09T21:14:39Z",
    },
  }],
  kind: "List",
  metadata: {
    resourceVersion: "",
  },
}
```

### Test Plan

The new encoder will include robust tests to prove both its own output format
and its compatibility with standard YAML and JSON.

[X] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

None.

##### Unit tests

We will add unit tests for things like `omitempty` and `omitzero`, for strings
that YAML coerces into non-string types, self-marshaling, and so forth.

We will include tests which check specific input constructions against known

> output, to ensure the subtleties of the specified syntax are correct. . We will

Reviewer 9: Ditto.

unit test the kubectl format option.

We will also include a round-trip fuzz-test which populates a complex structure
(which exercises as many edge-cases as possible) with random data, renders that
to KYAML, parses that KYAML back into the original type (with the standard YAML
decoder), and compares the original and result objects. Doing this enough times
should give us fairly high confidence.

##### Integration tests

N/A

##### e2e tests

N/A

### Graduation Criteria

This KEP introduces a new env-var feature gate to prevent premature
dependencies from forming.

#### Alpha

The new env-var feature gate will be checked before allowing use of the "kyaml"
output option. If `$KUBECTL_KYAML` is not set to "true", the "kyaml" output
option will not be available.

- The `kubectl` output format `kyaml` is implemented
- The gate is disabled by default
- All unit tests are complete
- Gather feedback from developers and surveys

#### Beta

The new env-var feature gate is enabled by default, but still disableable. If
`$KUBECTL_KYAML` is not set to "false", the "kyaml" output option will be
available.

> - Resolve any identified issues.

Reviewer 9: This still holds.

> - Resolve any identified issues.

Author: dropped

Taken together,
these mean that KYAML is not whitespace-sensitive,
which makes templating and other text-patching much easier.
YAML calls this “flow style”,
as opposed to the conventional “block style”.

This might sound a lot like JSON,
because it is!
Remember that JSON is valid YAML.
KYAML is friendlier than JSON for humans
to read,
to write,
and to commit to source control.
Unlike JSON, KYAML:
1. Allows comments (when authored by users).
2. Allows trailing commas on the last item in a list or struct/map.
3. Does not require quoted keys.

KYAML also includes a “header” (still valid YAML)
which helps to disambiguate a KYAML document from an ill-formed JSON document
(see: [kubernetes/kubernetes#77946](https://github.com/kubernetes/kubernetes/issues/78946)).

KYAML also tries to economize on vertical space by ["cuddling"](https://blog.gskinner.com/archives/2008/11/curly_braces_to.html) some kinds of brackets together.

Because KYAML is YAML, a KYAML multi-document is a YAML multi-document.

To make this change, we will:
1. Create new code (see details below) which serializes any object into KYAML.

> 2. Introduce a kubectl output format, `-o=kyaml`, which uses this new package.

Reviewer 9: > Will add as an unresolved - is there a sig-CLI convention? @soltysh ??

For posterity, since I already spoke with Tim about it on slack. Yes kubectl have their own feature gates mechanism, through environment variables. You define them [here](https://github.com/kubernetes/kubernetes/blob/8a0f6370e4b53b648050c534f0ee11b776f900a6/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L427-L458) and [here's sample usage](https://github.com/kubernetes/kubernetes/blob/8a0f6370e4b53b648050c534f0ee11b776f900a6/staging/src/k8s.io/kubectl/pkg/cmd/cmd.go#L362-L364).

> 3. Create tooling which reformats YAML or JSON as KYAML (see details below).

Reviewer: yamlft is the right idea but we don't release it currently and we discourage importing from k/k, we might consider options to stick a tool somewhere more obviously reusable later

I think k/k's yamlfmt is the place to start exploring this though, and we can follow with identifying where we can publish it or something like it

> 3. Create tooling which reformats YAML or JSON as KYAML (see details below).

Reviewer 9: Maybe `yamlfmt` should become part of `sigs.k8s.io/yaml` library along with the new Marshaler for kyaml?

> 3. Create tooling which reformats YAML or JSON as KYAML (see details below).

Author: That's an option - @liggitt?

> 3. Create tooling which reformats YAML or JSON as KYAML (see details below).

Reviewer: I thought about this, but we'd have to be really careful to avoid bringing in CLI dependencies.
We could also adding yamlfmt to k/k releases.

> 3. Create tooling which reformats YAML or JSON as KYAML (see details below).

Reviewer: I think it might be best IF we want to move it to k-sigs to move it to it's own repo as a thin CLI wrapper around sigs.k8s.io/yaml, but then we have to manage releasing etc seperately.

4. Work with SIG Docs to update examples to this format, both example command-lines and examples of API objects.
5. Look for other examples across kubernetes-affiliated git repositories, and offer PRs to convert to KYAML.
6. Make a lot of noise about it, and try to get more people to use it

The rationale for changing everything everywhere is simple.
Many users start with examples,
which they copy and modify.
By changing the examples we seed the change in the ecosystem.
One success metric might be
that KYAML becomes the preferred format in the most common helm charts.

### Notes/Constraints/Caveats (Optional)

Important note:
This is EXCLUSIVELY about client-side formatting.
This KEP *does not* propose any changes to apiserver
or other components besides kubectl.

Syntax has always been a magnet for arguments in our industry.
We expect this to be no exception.
Reasonable people will disagree with our choices,
which is fine, but we should hold a strong opinion on this.

YAML is a very large and complex specification.
We expect that we will find places where KYAML may be less readable than conventional YAML.
We should revisit this specification when that happens.

> ### Risks and Mitigations

Reviewer 9: ```
[Drawbacks](#drawbacks)
```

the server.

One side-effect of this is that we cannot reformat some valid YAML documents
into KYAML, because there are things that YAML allows that JSON does not, such
as compound non-string keys, and things that Go's JSON encoder quietly converts
to strings, such as numbers.

Today we have forked go-yaml.v2 AND go-yaml.v3 into sigs.k8s.io/yaml.  While
KYAML is an output format, our tooling goals require that we get access to a
YAML AST.  Fortunately, go-yaml.v3 offers a way to decode YAML into such an
AST. If we ever change YAML libraries (e.g.
[https://github.com/goccy/go-yaml](https://github.com/goccy/go-yaml)), we will
need to rewrite this code.

### Comments

When reformating existing YAML with comments, we will attempt to place comments
in "obvious" places.  For example, a line-comment on a scalar field is
obviously placed at the end of the line.  These cases will not be called out in
the subsequent sections of this KEP.  In some cases, the comment placement
may not be obvious, and so will be called out explicitly.

### Header

At the top of each KYAML document (or multi-document) will be the following header:

> ```

Reviewer: Let's do `# YAML`, it close to the standard and yaml.org is consistently all caps

> ```

Author: In next push.

type-ambiguous (e.g. `no`)

Values are rendered as per their type.

<<[UNRESOLVED quoted strings]>>
Should we treat strings for field names and map keys:
- As described above, with label keys quoted but field names not
- Allow unquoted label keys and more complex (but not ambiguous) values
- Always quote strings, including ALL map keys and field names
<<[/UNRESOLVED>>

##### Lists

Slices and arrays are rendered with square-braces. If a struct has a
line-comment it will be rendered after the opening brace.

List elements are rendered as per their type.

##### Pointers

Pointers are rendered as `null` or dereferenced and rendered as per the
pointee’s type.

##### Self marshalers

Types which implement the Go standard `json.Marshaler` interface
(`MarshalJSON()`) are rendered by marshaling to JSON.

Types which implement the `yaml.Marshaler` interface (`MarshalYAML()`) are
rendered as per their type.  `MarshalYAML()` is not a Go standard and is not
used by KYAML.

##### Interfaces

Interfaces are rendered as `null` or their true value type.

##### Anything else

Any other type (as per `reflect.Value.Kind()`) will produce an error.

> #### Multi-line strings

Reviewer 4: > It would be neat if you could tell a pretty printer: hey, this multi line string is JSON, please also pretty-print that.

I'm also not quite sure what you mean by this ... I don't think we should do anything that would take a YAML string and treat it as structured data generically. That kind of non-standard interpretation is a good way to worm things through layers that interpret the data differently.

comma is not required, but KYAML always emits it.

Example:

```
labels: {
  app: "foobar",
  foo: "bar",
  something: “12345”,
},
```

### Braces

Paired braces are always cuddled, unless the next item has a head-comment or
the previous item has a foot-comment.

This is best illustrated with an example:

Cuddled:
```
list: [{
  foo: “abc”, # This is a line-comment
  bar: 123,
}, {
  foo: “def”,
  bar: 456,
}]
```

Partially cuddled, with a head-comment.
```
list: [
# This is a head-comment
{
  foo: “abc”,
  bar: 123,
}, {
  foo: “def”,
  bar: 456,
}]
```

Partially cuddled, with a foot-comment.
```
list: [{
  foo: “abc”,
  bar: 123,
},
# This is a foot comment
{
  foo: “def”,
  bar: 456,
}]
```

<<[UNRESOLVED uncuddle if comments]>>
The "mixed" cuddling demonstrated above is a little odd.  As an alternative, we
could detect if any item has head- or foot-comments and just uncuddle all
braces in that list.

Or we could style lists with an extra indent like:
```
list: [{
    foo: “abc”,
    bar: 123,
  },
  # This is a foot comment
  {
    foo: “def”,
    bar: 456,
  }]
```
<</UNRESOLVED>>

### Indents

Indents follow YAML’s two-space convention.

> <<[UNRESOLVED readability]>>

Author: I'd rather go for few or even no options, but I could be arm-twisted.  This feels like gravy on top of an already sprawling KEP :)

> <<[UNRESOLVED readability]>>

Reviewer 4: > I'd rather go for few or even no options, but I could be arm-twisted.

+1, please no options

> <<[UNRESOLVED readability]>>

Author: Removing this - it doesn't seem like anyone is cheering for wider indents in KYAML?

        type: "PodScheduled",
      }],
      containerStatuses: [{
        allocatedResources: {
          cpu: "100m",
        },
        containerID: "containerd://49d6e2a850d3569145de92a1138b377a904954b48b23905af00b15ec5393de77",
        image: "k8s.gcr.io/serve_hostname:v1.4",
        imageID: "sha256:b56f22170982bc46dee99f7157be69dce727d3b89ee5698d15fda7a14a124200",
        lastState: {},
        name: "serve-hostname-xqnl5",
        ready: true,
        resources: {
          requests: {
            cpu: "100m",
          },
        },
        restartCount: 0,
        started: true,
        state: {
          running: {
            startedAt: "2025-05-09T21:14:41Z",
          },
        },
        volumeMounts: [{
          mountPath: "/var/run/secrets/kubernetes.io/serviceaccount",
          name: "kube-api-access-g57rz",
          readOnly: true,
          recursiveReadOnly: "Disabled",
        }],
      }],
      hostIP: "10.40.0.9",
      hostIPs: [{
        ip: "10.40.0.9",
      }],
      phase: "Running",
      podIP: "10.64.3.4",
      podIPs: [{
        ip: "10.64.3.4",
      }],
      qosClass: "Burstable",
      startTime: "2025-05-09T21:14:39Z",
    },
  }],

>   kind: "List",

Author: That's a whole different KEP :)

Unfortunately, kubectl converts everything into Unstructured which is a map, and loses field ordering.  KYAML should be no different than any other format in this regard.

Those will be handled as bugs.

###### What steps should be taken if SLOs are not being met to determine the problem?

N/A

## Implementation History

May 09, 2025: KEP draft v0

## Drawbacks

Having a local dialect of YAML could be confusing to people who are used to
conventional YAML.

Trying to text-patch KYAML with plain YAML (as in a helm chart) will almost
certainly not work (mixing flow- and block-styles is a mess).

More serialization “languages” dilutes knowledge.

## Alternatives

We could just use JSON and deal with it.

We could tell users to use JSON with comments and a JSON minifier tool to strip
those out before feeding to kubernetes.

We could “ban” the use of raw JSON (except to the API) and instead use
JSON-with-comments (via YAML parsing) in all configuration logic.  This doesn’t
solve for kubectl output.

We could adopt [HJSON](https://hjson.github.io/) or
[JSONC](https://www.npmjs.com/package/jsonc) or [JSON5](https://json5.org/).
These would require choosing between them, taking new dependencies, and
learning new grammars.  All of them are very similar to KYAML, which has none
of those drawbacks.

> 

Author: I'm going to leave that for a followup KEP :)

> 

Reviewer: This is fragile to human edits while KYAML generally isn't and there's probably 2000 bikesheds to paint to decide what these comments would look like. +1 for future KEP instead

6. Make a lot of noise about it, and try to get more people to use it

The rationale for changing everything everywhere is simple.
Many users start with examples,
which they copy and modify.
By changing the examples we seed the change in the ecosystem.
One success metric might be
that KYAML becomes the preferred format in the most common helm charts.

<<[UNRESOLVED to gate or not to gate]>>
Should we require an env var to enable kyaml in the first release?  Are there
sig-cli conventions here that I don't know?
<</UNRESOLVED>>

### Notes/Constraints/Caveats

**Important note:**
This is EXCLUSIVELY about client-side formatting.
This KEP *does not* propose any changes to apiserver
or other components besides kubectl.

Syntax has always been a magnet for arguments in our industry.
We expect this to be no exception.
Reasonable people will disagree with our choices,
which is fine, but we should hold a strong opinion on this.

YAML is a very large and complex specification.
We expect that we will find places where KYAML may be less readable than conventional YAML.
We should revisit this specification when that happens.
Givent that KYAML is YAML it should be possible to revise

> without breaking anyone.

Reviewer 4: Any byte-accurate tests consuming kyaml output will break (we saw this happen on yamlv2 to yamlv3 indent / wrapping changes). If kyaml output is not guaranteed to be byte-stable, we need to clearly / loudly state this, and ideally, do something in the implementation so someone depending on it being byte-stable notices (e.g. make some aspect of the output just *slightly* non-deterministic or something)

Author: Thoughts on how to prevent Hyrum's law?  I don't think we want to (or can, really) randomize field ordering, for example.  Is it sufficient to say "this is a bad idea" ?  Do we need to version KYAML?

Author: Added words.

Reviewer 2: Let's not make the output non-deterministic. The "dump diff between two objects" usage would then show semantically irrelevant differences.

Reviewer: I think it's unreasonable to depend on a local formatter being byte-stable between minor releases.
If the server were serving it, maybe, but that functionality is explicitly not included in this KEP.

Reviewer 2: My use case is for invoking the formatter twice within the same process.

I think we all agree that no stability should be expected between releases, it's just a question of how to make users aware of that.

obviously reusable place for it.

A more ecosystem-friendly solution is to write a new function or package under
sigs.k8s.io/yaml](sigs.k8s.io/yaml) which can be used by other tools.

<<[/UNRESOLVED]>>

### Tooling

k/k has a CI verifier (hack/verify-yamlfmt.sh) which looks for YAML files and
ensures they are formatted “correctly”. The only meaning of “correct” is “block
style”, and it corrupts KYAML input today (aside: it also does not handle
multi-documents).  If we intend to convert examples, we need to fix this.

We propose to add flags to it which specify to use “conventional” YAML or KYAML
output, and set the default to KYAML (or to change the parent script to always
specify KYAML).

This tooling will preserve comments as best it can. Unfortunately, go-yaml (our
YAML library) [does not always handle comments properly](https://github.com/yaml/go-yaml/issues/8).
It is possible that some comments will be formatted wrongly, or lost entirely.
The affected comments are not representable in YAML block-style, so this is not
exactly a regression.

A guiding principle for the tooling is that whatever output the tool produces,

> when fed back into the tool, should not be modified again.

Reviewer 4: nit: figure out how to make this true while also not allowing people to capture KYAML fixtures and use them in tests that will break if we later tweak KYAML rendering

#### How?

To ensure maximum fidelity to JSON, KYAML rendering of Go objects will first
marshal them to JSON, then render the JSON as KYAML.  This process means we get
all of the JSON tag logic for "free", and that we can use the same code to
render Go objects and re-render YAML or JSON input. That extra conversion is
ugly, but this is not generally performance-sensitive code and is not run on
the server.

One side-effect of this is that we cannot reformat some valid YAML documents
into KYAML, because there are things that YAML allows that JSON does not, such
as compound non-string keys, and things that Go's JSON encoder quietly converts
to strings, such as numbers.

As of kubernetes 1.33, we have forked go-yaml.v2 AND go-yaml.v3 into
sigs.k8s.io/yaml.  While KYAML is an output format, our tooling goals require
that we get access to a YAML AST.  Fortunately, go-yaml.v3 offers a way to
decode YAML into such an AST. If we ever change YAML libraries (e.g.
[https://github.com/goccy/go-yaml](https://github.com/goccy/go-yaml)), we will
need to rewrite this code.

### Comments

> When reformating existing YAML with comments, we will attempt to place comments

Reviewer 4: ```suggestion
When reformatting existing YAML with comments, we will attempt to place comments
```

in "obvious" places.  For example, a line-comment on a scalar field is
obviously placed at the end of the line.  These cases will not be called out in
the subsequent sections of this KEP.  In some cases, the comment placement
may not be obvious, and so will be called out explicitly.

### Header

At the top of each KYAML document (or multi-document) will be the following header:

```
# YAML
---
```

This disambiguates it from JSON, since both KYAML and JSON begin with an
opening curly-brace.  Kubernetes versions 1.33 and higher can handle KYAML
without the header, but it is needed for versions before that.

### Per-type rendering

#### Scalars

Integers and floats are rendered as their natural numeric representation.

Simple strings are rendered as double-quoted and escaped. Multi-line strings
require special consideration, see below.

Booleans are rendered as `true` or `false`.

#### Mappings (structs and maps)

##### Structs

Structs are rendered with curly-braces, and each field on a line of its own. If
a struct has a line-comment it will be rendered after the closing brace (since

> this is the only place it will be found upon re-reading).

Reviewer 4: wow... that's such an unfortunate location to have to put the comment :-/

That means comments like this are lost:
```yaml
# comment
foo:
- ...
```

(asking just for clarity, not suggesting this is worth a ton more effort to fix right now)

Author: In "block style", that comment is a "head comment":

```
$ go run ./yaml_renderer.go << EOF
> # comment
> foo: bar
> EOF
# YAML
---
{
  # comment
  foo: "bar",
}
```

The problem is more like this:

```
$ go run ./yaml_renderer.go << EOF
> { # this comment will be lost
>   foo: bar # line comment
> } # this comment will be kept
> EOF
# YAML
---
{
  foo: "bar", # line comment
} # this comment will be kept
```

Reviewer 6: The dropped comment there is just a bug in go-yaml, that can be fix very soon.

Field names are not quoted unless they are not obviously strings (which

> includes prefixed label keys) or are one of the known type-ambiguous words

Author: I think we should allow labels unquoted.  will clarify

(e.g. `no`).  Field values are rendered as per their type.  Embedded structs
will be rendered as inline fields.

KYAML does not guarantee that fields will be rendered in the order they are
defined in the struct.  Implementations may preserve field order, but `kubectl`
itself does not guarantee this.  For example, `kubectl` converts everything to an
unstructured type internally, which loses the field ordering.

Fields which have a `json` struct tag will be rendered according to that tag,
by nature of the fact that we marshal Go objects to JSON first.

The `yaml` struct tag will be ignored.

##### Maps

Maps are rendered with curly-braces, and each key on a line of its own. If a

> map has a line-comment it will be rendered after the opening brace.

Author: oops! Should have update this.

restrictive than plain YAML. Reformatting YAML with non-string keys into KYAML
will fail.

Keys are sorted and are not quoted unless they are non-trivial or are
type-ambiguous (e.g. `no`)

Values are rendered as per their type.

<<[UNRESOLVED quoted strings]>>
Should we treat strings for field names and map keys:
- As described above, with label keys quoted but field names not
- Allow unquoted label keys and more complex (but not ambiguous) values
- Always quote strings, including ALL map keys and field names
<<[/UNRESOLVED>>

##### Lists

Slices and arrays are rendered with square-braces. If a struct has a
line-comment it will be rendered after the opening brace.

List elements are rendered as per their type.

##### Pointers

Pointers are rendered as `null` or dereferenced and rendered as per the
pointee’s type.

##### Self marshalers

Types which implement the Go standard `json.Marshaler` interface
(`MarshalJSON()`) are rendered by marshaling to JSON.

Types which implement the `yaml.Marshaler` interface (`MarshalYAML()`) are
rendered as per their type.  `MarshalYAML()` is not a Go standard and is not
used by KYAML.

> 

Author: Added under "tooling"

"""
Complex YAML input, with things like anchors, aliases, explicit tags (!!...),                                                                                          
and global tags (%TAG !...) may be simplified when rendered to KYAML, as long
as the resulting unmarshaled object would be identical.
"""

> 

Reviewer 6: fwiw `%YAML` and `%TAG` lines in YAML are called directives. "global tags" isn't a thing.

KYAML also tries to economize on vertical space by ["cuddling"](https://blog.gskinner.com/archives/2008/11/curly_braces_to.html) some kinds of brackets together.

Because KYAML is YAML, a KYAML multi-document is a YAML multi-document.

To make this change, we will:
1. Create new code (see details below) which serializes any object into KYAML.
2. Introduce a kubectl output format, `-o=kyaml`, which uses this new package.
3. Create tooling which reformats YAML or JSON as KYAML (see details below).
4. Work with SIG Docs to update examples to this format, both example command-lines and examples of API objects.
5. Look for other examples across kubernetes-affiliated git repositories, and offer PRs to convert to KYAML.
6. Make a lot of noise about it, and try to get more people to use it

The rationale for changing everything everywhere is simple.
Many users start with examples,
which they copy and modify.
By changing the examples we seed the change in the ecosystem.
One success metric might be
that KYAML becomes the preferred format in the most common helm charts.

<<[UNRESOLVED to gate or not to gate]>>

> Should we require an env var to enable kyaml in the first release?  Are there

Reviewer 10: Does kubeadm also need to honor this environment variable and output kyaml formatted static Pod manifests along with kyaml formatted configuration files?

Reviewer 10: /cc @neolit123

Reviewer 11: > Does kubeadm also need to honor this environment variable and output kyaml formatted static Pod manifests along with kyaml formatted configuration files?

>> This KEP *does not* propose any changes to apiserver
or other components besides kubectl.

kubeadm also can output 'config' for users, and it utilizes k8s.io/cli-runtime/pkg/printers and k8s.io/cli-runtime/pkg/genericclioptions. if kyaml support is integrated there for kubectl support, presumably, kubeadm will end up supporting it too.

Author: Thanks!  Will note.

Syntax has always been a magnet for arguments in our industry.
We expect this to be no exception.
Reasonable people will disagree with our choices,
which is fine, but we should hold a strong opinion on this.

YAML is a very large and complex specification.
We expect that we will find places where KYAML may be less readable than conventional YAML.
We should revisit this specification when that happens.
Given that KYAML is YAML it should be possible to revise
without breaking anyone.

### Risks and Mitigations

Risk: Our encoder produces some output which is not valid YAML.
Mitigation: That’s a bug. It’s hard to imagine a case that can’t be fixed.

Risk: The result could be net worse than plain YAML.
Mitigation: We will produce many examples and include them in the review process. We will solicit feedback from various groups on the result.

Risk: This gives a false sense of safety, but doesn’t SOLVE  the problems with YAML, just avoids them.
Mitigation: Docs?

Risk: Like all APIs, people will come to depend on the details, which may prevent us from changing them.
Mitigation: Be as precise as possible where we are confident, but leave room where we are not.

> ## Design Details

Reviewer 10: Will we have a dedicated document to formally record the format specifications for kyaml? Or perhaps a website where we can document revisions to kyaml (if we need to update it in the future), similar to: https://yaml.org/

While I’ve never paid attention to yaml.org before and have always written YAML files based on intuition or by referencing others’ formats, kyaml, but as a subset of YAML, appears to have stricter constraints.

Author: This spec is primarily about OUTPUT.  Any valid YAML is still allowed, I am just trying to restrict our output format to something more explicit.

Yes, once we pin a spec down, I am happy to extract it to a different place.

Author: I'm of two minds on this.  On one hand, I don't mind writing it down (as we did here).  On the other hand, who benefits from it, really?  Anyone consuming KYAML should have a full YAML parser, and it really doesn't matter if tools follow the spec strictly for their own output because it is still YAML.

If someone wants to emit the same format from their tool, that's fine.  I don't want to say that we can never ever change the details of this, though.  As long as KYAML is YAML, minor formatting changesd are (should be) inconsequential.  E.g. if we changed indent size, or offered a "compact" render mode, or changed cuddling - it shouldn't be a dramatic thing.

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

NO

### Troubleshooting

If someone thinks the KYAML rendering is wrong they can instead render to plain
YAML or JSON and compare.  Additionally, they can render to KYAML and then use
yamlfmt to convert that to plain YAML.

###### How does this feature react if the API server and/or etcd is unavailable?

N/A

###### What are other known failure modes?

It’s possible we missed some esoteric structure that renders incorrectly.
Those will be handled as bugs.

###### What steps should be taken if SLOs are not being met to determine the problem?

N/A

## Implementation History

May 09, 2025: KEP draft v0

## Drawbacks

Having a local dialect of YAML could be confusing to people who are used to
conventional YAML.

Trying to text-patch KYAML with plain YAML (as in a helm chart) will almost
certainly not work (mixing flow- and block-styles is a mess).

More serialization “languages” dilutes knowledge.

[Yet another standard](https://xkcd.com/927/). Rebuttal: KYAML is YAML and
users can provide vanilla YAML anyywhere that can accept KYAML.

## Alternatives

> We could just use JSON and deal with it.
> 
> We could tell users to use JSON with comments and a JSON minifier tool to strip
> those out before feeding to kubernetes.
> 
> We could “ban” the use of raw JSON (except to the API) and instead use
> JSON-with-comments (via YAML parsing) in all configuration logic.  This doesn’t
> solve for kubectl output.
> 
> We could adopt [HJSON](https://hjson.github.io/) or
> [JSONC](https://www.npmjs.com/package/jsonc) or [JSON5](https://json5.org/).
> These would require choosing between them, taking new dependencies, and
> learning new grammars.  All of them are very similar to KYAML, which has none
> of those drawbacks.

Reviewer 12: ```suggestion
1. We could just use JSON and deal with it.

1. We could tell users to use JSON with comments and a JSON minifier tool
   to strip those comments out before feeding to Kubernetes.

1. We could “ban” the use of raw JSON (except to the API) and instead use
   JSON-with-comments (via YAML parsing) in all configuration logic.
   This doesn’t solve for kubectl output.

1. We could adopt one of:
   - [HJSON](https://hjson.github.io/)
   - [JSONC](https://www.npmjs.com/package/jsonc)
   - [JSON5](https://json5.org/)

  These would require choosing between them, taking new dependencies, and
  learning new grammars.
  All of them are very similar to KYAML, which has none of those drawbacks.
  KYAML is not intended to be a standard that other projects adopt; for interoperability,
  we'll offer YAML, JSON and (on the wire) CBOR.
```

We could us KYAML on the wire, but we long-ago decided NOT to support YAML on

> server-side.

Reviewer 12: Actually, kube-apiserver will happily accept or serve YAML if you negotiate it: https://kubernetes.io/docs/reference/using-api/api-concepts/#yaml-encoding

Author: @liggitt is this right?  I thought we explicitly DON'T want the server parsing YAML?

Author: Oof, I was mistaken.  I do not think we should do anything special for KYAML.

Mitigation: Be as precise as possible where we are confident, but leave room where we are not.

## Design Details

Creating an encoder for this format is not very technically challenging,
but there are many details to consider.

### Future compatibility

KYAML is, and will remain, a strict subset of YAML.  Compliant YAML parsers
should always be able to parse KYAML data and we will endeavour to make sure
they can all retain fidelity with comments.

It may be that, over time, we find some aspects of KYAML's exact rendering that
can be inmproved, and we will consider those opportunities. The implication of
this is that clients which rely on the byte-exact results of KYAML's rendering
might be broken by future changes. We will try to avoid this, but we will not
guarantee it.

### Packaging

<<[UNRESOLVED where the code lives]>>

The simplest thing to do is to write a new implementation of
k8s.io/cli-runtime/pkg/printers.ResourcePrinter (in
staging/src/k8s.io/cli-runtime/pkg/printers). However, this isn’t the most
obviously reusable place for it.

A more ecosystem-friendly solution is to write a new function or package under

> sigs.k8s.io/yaml](sigs.k8s.io/yaml) which can be used by other tools.

Reviewer 2: >  >    Where the code lives
>
> If I hear no news I will create a kyaml pkg under sigs.k8s.io/yaml for the rendering logic and put the printer under staging/src/k8s.io/cli-runtime/pkg/printers using the sigs renderer.

I still think that putting the rendering logic into a separate module with no dependencies other than the standard library is desirable. Is that option definitely off the table because it would make the implementation harder?

Reviewer 2: If you want a single git repo with the entire source, then sigs.k8s.io/yaml could become a [multi-module repo](https://go.dev/doc/modules/managing-source#multiple-module-source).

Reviewer: There are only 4 of the 31 staging modules that do not depend on sigs.k8s.io/yaml currently.
Of those only one imports klog.

```console
$ git describe
v1.34.0-alpha.0-705-g148761ce2c3

$ grep 'sigs.k8s.io/yaml' staging/src/k8s.io/*/go.mod
staging/src/k8s.io/api/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/apiextensions-apiserver/go.mod:	sigs.k8s.io/yaml v1.4.0
staging/src/k8s.io/apimachinery/go.mod:	sigs.k8s.io/yaml v1.4.0
staging/src/k8s.io/apiserver/go.mod:	sigs.k8s.io/yaml v1.4.0
staging/src/k8s.io/cli-runtime/go.mod:	sigs.k8s.io/yaml v1.4.0
staging/src/k8s.io/client-go/go.mod:	sigs.k8s.io/yaml v1.4.0
staging/src/k8s.io/cloud-provider/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/cluster-bootstrap/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/code-generator/go.mod:	sigs.k8s.io/yaml v1.4.0
staging/src/k8s.io/component-base/go.mod:	sigs.k8s.io/yaml v1.4.0
staging/src/k8s.io/component-helpers/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/controller-manager/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/cri-client/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/csi-translation-lib/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/dynamic-resource-allocation/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/endpointslice/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/kube-aggregator/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/kube-controller-manager/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/kube-proxy/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/kube-scheduler/go.mod:	sigs.k8s.io/yaml v1.4.0
staging/src/k8s.io/kubectl/go.mod:	sigs.k8s.io/yaml v1.4.0
staging/src/k8s.io/kubelet/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/metrics/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/pod-security-admission/go.mod:	sigs.k8s.io/yaml v1.4.0
staging/src/k8s.io/sample-apiserver/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/sample-cli-plugin/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect
staging/src/k8s.io/sample-controller/go.mod:	sigs.k8s.io/yaml v1.4.0 // indirect

$ find staging/src/k8s.io/*/go.mod | wc -l
      31

$ grep 'sigs.k8s.io/yaml' staging/src/k8s.io/*/go.mod | wc -l
      27

$ diff <(grep -l 'sigs.k8s.io/yaml' staging/src/k8s.io/*/go.mod | sort) <(find staging/src/k8s.io/*/go.mod | sort)
12a13
> staging/src/k8s.io/cri-api/go.mod
16a18,19
> staging/src/k8s.io/externaljwt/go.mod
> staging/src/k8s.io/kms/go.mod
23a27
> staging/src/k8s.io/mount-utils/go.mod

$ grep klog staging/src/k8s.io/cri-api/go.mod staging/src/k8s.io/externaljwt/go.mod staging/src/k8s.io/kms/go.mod staging/src/k8s.io/mount-utils/go.mod
staging/src/k8s.io/mount-utils/go.mod:	k8s.io/klog/v2 v2.130.1
```

Author: It feels like a lot of work (setting up a repo) for low reward, to me.

Additionally, I need to import `sigs.k8s.io/yaml` so I can use `yaml.Node` .

Author: It should not be difficult to write a version that renders kyaml from a Go object without importing sigs.k8s.io/yaml but I think it would be a different implementation.  

I *have* that implementation, but I wasn't excited about maintaining two of them over time :)

FWIW - it's pretty similar to the rendering code in `go-logr/logr/funcr` - I could maybe be convinced to put it there once the dust settles on the exact spec.

- [ ] (R) KEP approvers have approved the KEP status as `implementable`
- [ ] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
  - [ ] e2e Tests for all Beta API Operations (endpoints)
  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]

> - [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

Author: Does everyone think "implementable is OK with this spec" ?

Is everyone OK with test plan?

(aka  [“The Norway Bug”](https://hitchdev.com/strictyaml/why/implicit-typing-removed/)),
`_42`, `_4_2_`, and `__________4__________2__________` are all treated as numbers,
and `11:00` is also treated as a number (base 60).

As an alternative to YAML most of Kubernetes also supports JSON,
But JSON has its own drawbacks.
It does not support comments.
It is very particular about trailing commas.
It requires quoted keys.

In addition to `kubectl` output,
most Kubernetes config files are either YAML or JSON.
JSON lacks comments,
which are very valuable for configuration,
so most configuration and tooling is built around YAML.

The most frustrating part of YAML is that it doesn’t have to be this bad.
YAML is a very broad specification.
Inside that specification is a reasonable grammar screaming to be let out.

If we change our tooling and examples to use a less error-prone dialect of YAML,
we can encourage the ecosystem to avoid some of the YAML pitfalls
and reduce common mistakes and frustration.

### Goals

1. Specify a YAML dialect which is 100% compatible with existing parsers and tooling.
2. Provide libraries and tooling for encoding and converting to this formatting.

> 3. Provide a KYAML output format for kubectl

Reviewer 9: ```suggestion
3. Provide a KYAML output format for kubectl.
```

> 4. Update examples and documentation throughout the project

Reviewer 9: ```suggestion
4. Update examples and documentation throughout the project.
```

5. Address common YAML pitfalls through an opinionated formatting approach.

> 6. Documenting the design and reasons for KYAML and the 100% compatibility with tooling expecting arbitrary YAML.

Reviewer 9: ```suggestion
6. Document the design and reasons for KYAML and the 100% compatibility with tooling expecting arbitrary YAML.
```

### Non-Goals

> 1. Introducing alternative configuration languages that are not compatible with existing tooling.

Reviewer 9: ```suggestion
1. Introduce alternative configuration languages that are not compatible with existing tooling.
```

> 2. Server-side support for KYAML or YAML

Reviewer 9: ```suggestion
2. Server-side support for KYAML.
```
@BenTheElder pointed out [here](https://github.com/kubernetes/enhancements/pull/5296#discussion_r2105184539) we already support yaml in the apiserver.

> 3. Requiring projects to migrate by default.

Reviewer 9: ```suggestion
3. Require projects to migrate to by default.
```
Also which projects and migrate to what? I assume it's about non-k8s owned (non-k8s gh org)? Whereas the latter is about kyaml?

Author: We are not going to require kubeadm to migrate to kyaml, nor helm or any others.

As long as the result is still valid YAML.

KYAML addresses the problems cited above by:
1. Always quoting _value_ strings (no ambiguity around things like "NO").
2. Allowing (but not requiring) quoted keys, but always quoting them if they are not obviously safe (e.g. "no" would always be quoted).
3. Always using {} for structs and maps (no more obscure errors about mapping values).
4. Always using [] for lists (no more trying to figure out if a dash changes the meaning).

Taken together,
these mean that KYAML is not whitespace-sensitive,
which makes templating and other text-patching much easier.
YAML calls this “flow style”,
as opposed to the conventional “block style”.

This might sound a lot like JSON,
because it is!
Remember that JSON is valid YAML.
KYAML is friendlier than JSON for humans
to read,
to write,
and to commit to source control.
Unlike JSON, KYAML:
1. Allows comments (when authored by users).
2. Allows trailing commas on the last item in a list or struct/map.
3. Does not require quoted keys.

KYAML also includes a “header” (still valid YAML)
which helps to disambiguate a KYAML document from an ill-formed JSON document

> (see: [kubernetes/kubernetes#77946](https://github.com/kubernetes/kubernetes/issues/78946)).

Reviewer 9: ```suggestion
(see: [kubernetes/kubernetes#78946](https://github.com/kubernetes/kubernetes/issues/78946)).
```

### Notes/Constraints/Caveats

**Important note:**
This is EXCLUSIVELY about client-side formatting.
This KEP *does not* propose any changes to apiserver
or other components besides kubectl.

Syntax has always been a magnet for arguments in our industry.
We expect this to be no exception.
Reasonable people will disagree with our choices,
which is fine, but we should hold a strong opinion on this.

YAML is a very large and complex specification.
We expect that we will find places where KYAML may be less readable than conventional YAML.
We should revisit this specification when that happens.
Given that KYAML is YAML it should be possible to revise
without breaking anyone.

### Risks and Mitigations

Risk: Our encoder produces some output which is not valid YAML.
Mitigation: That’s a bug. It’s hard to imagine a case that can’t be fixed.

Risk: The result could be net worse than plain YAML.
Mitigation: We will produce many examples and include them in the review process. We will solicit feedback from various groups on the result.

Risk: This gives a false sense of safety, but doesn’t SOLVE  the problems with YAML, just avoids them.

> Mitigation: Docs?

Reviewer 9: Maybe the right answer here should be that our goal is not to modify YAML format, but rather focus on a valid YAML sub-set, which resolves known problems, increasing the success of our users.

Author: I took your words, thanks.

Risk: Like all APIs, people will come to depend on the details, which may prevent us from changing them.
Mitigation: Be as precise as possible where we are confident, but leave room where we are not.

## Design Details

Creating an encoder for this format is not very technically challenging,
but there are many details to consider.

### Feature gate

Literal feature gates are not super useful for CLI features, so SIG CLI has a
similar mechanism to gate features via environment variables.  We will use a
variable named `"KUBECTL_KYAML"` to gate this feature.

The output flags for JSON and YAML are handled in
`k8s.io/cli-runtime/pkg/genericclioptions`.  Unfortunately, that package is
already transitively depended on by `k8s.io/kubectl/pkg/cmd/util` where the CLI
gates are defined. Rather than trying to break the import cycle, we will just

> check the environment variable directly, without the utils package.

Reviewer 9: This is reasonable for now, even if you just copy&paste the similar functions. In the long run I'm planning to rewrite the mechanism we currently have and ideally re-use the one which is part of component-base package.

Author: ACK.  It's simple enough :)

### Future compatibility

KYAML is, and will remain, a strict subset of YAML.  Compliant YAML parsers
should always be able to parse KYAML data and we will endeavour to make sure
they can all retain fidelity with comments.

It may be that, over time, we find some aspects of KYAML's exact rendering that
can be inmproved, and we will consider those opportunities. The implication of
this is that clients which rely on the byte-exact results of KYAML's rendering
might be broken by future changes. We will try to avoid this, but we will not
guarantee it.

### Packaging

We will create a new `kyaml` package under `sigs.k8s.io/yaml` for the rendering
logic and put the printer under `staging/src/k8s.io/cli-runtime/pkg/printers`,
using the renderer.

### Components

The primary component affected by this KEP is `kubectl`, but any other
component which uses `k8s.io/cli-runtime/pkg/printers` and

> `k8s.io/cli-runtime/pkg/genericclioptions` will get KYAML support "for free".

Reviewer 9: When we created that `cli-runtime ` package we tried to promote this heavily to be used by all the kubectl plugins, I believe a lot of them do use it, which means they will also support the new format, and by that we should get even broader feedback surface.

style”, and it corrupts KYAML input today (aside: it also does not handle
multi-documents).  If we intend to convert examples, we need to fix this.

We propose to add flags to it which specify to use “conventional” YAML or KYAML
output, and set the default to KYAML (or to change the parent script to always
specify KYAML).

This tooling will preserve comments as best it can. Unfortunately, go-yaml (our
YAML library) [does not always handle comments properly](https://github.com/yaml/go-yaml/issues/8).
It is possible that some comments will be formatted wrongly, or lost entirely.
The affected comments are not representable in YAML block-style, so this is not
exactly a regression.

A guiding principle for the tooling is that whatever output the tool produces,
when fed back into the same version of the tool, should not be modified again.
As mentioned above, we MIGHT make formatting changes over time, so this
property may not hold across versions of the tooling.

Complex YAML input, with things like anchors, aliases, explicit tags (!!...),
and global tags (%TAG !...) may be simplified when rendered to KYAML, as long

> as the resulting unmarshaled object would be identical.

Reviewer 9: I'm not quite sure what you mean "maybe be simplified" here. Care to explain?

Author: YAML allows things like aliases which will be dereferenced when run through this:

e.g.

```
$ go run ./yaml_renderer.go 
- &flag Apple
- Beachball
- Cartoon
- Duckface
- *flag
---
[
  "Apple",
  "Beachball",
  "Cartoon",
  "Duckface",
  "Apple",
]
```

Author: Added note

- The gate is disabled by default
- All unit tests are complete
- Gather feedback from developers and surveys

#### Beta

The new env-var feature gate is enabled by default, but still disableable. If
`$KUBECTL_KYAML` is not set to "false", the "kyaml" output option will be
available.

- Resolve any identified issues.

#### GA

The new env-var feature gate is removed, the "kyaml" output option is always
available.

- A proper specification is published
- The majority of examples in the kubernetes GitHub orgs are converted to KYAML
- No new bug reports

This KEP does not propose to add conformance tests.

### Upgrade / Downgrade Strategy

This is a client-side feature.  Users should not use this in automation until
they are sure their clients are updated, but there is little risk of breakage
from ad hoc use.

KYAML is valid YAML. `kubectl` already accepts KYAML as defined here and with
arbitrary variations, as long as it remains valid YAML.

### Version Skew Strategy

N/A

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [ ] Other

>   - Opt-in, available without gates

Reviewer 9: ```
- [x] Other
  - Describe the mechanism: The environment variable `KUBECTL_KYAML=true`.
```
